### PR TITLE
[oscparse] improvements

### PR DESCRIPTION
**Pull request by porres**
[oscparse] always parses addresses as symbols, even if they look like floats, which make it really hard to use [route]. I propose a flag that turns addresses that look like floats into floats.

We also add an outlet that gives the index in the output list that splits the address from the data

closes https://github.com/pure-data/pure-data/issues/1299

also closes https://github.com/pure-data/pure-data/issues/1858

### CODE REVIEW

File: src/x_misc.c [hunk 1]
```diff
@@ -289,7 +290,8 @@ static t_class *oscparse_class;
 
 typedef struct _oscparse
 {
-    t_object x_obj;
+    t_object    x_obj;
+    t_int       x_flag;
```
Comment (Spacechild1) [inline src/x_misc.c:294]: `t_int` is meant for pointers. If you need an int, use `int` instead.

File: src/x_misc.c [hunk 2]
```diff
@@ -385,8 +387,20 @@ static void oscparse_list(t_oscparse *x, t_symbol *s, int argc, t_atom *argv)
...
         dataonset, nfield); */
     for (i = j = 0; i < typeonset-1 && argv[i].a_w.w_float != 0 &&
-        j < outc; j++)
-            SETSYMBOL(outv+j, grabstring(argc, argv, &i, 1));
+         j < outc; j++)
+    {
+        if(x->x_flag)
+        {
+            t_symbol *sym = grabstring(argc, argv, &i, 1);
+              t_float f = 0.0f;
```
Comment (Spacechild1) [inline src/x_misc.c:395]: watch your indentation ;-)

**Comment by Spacechild1** [comment]
could the flag be just `-f` instead of `-float`?

**Comment by porres** [comment]
I thought '-f' would be confusing as it is used for [oscformat] but it has a different meaning there ('format').

Note that [oscparse] has no arguments, so maybe we could just have an argument instead.

 I also think there could be a message method to change this behaviour as well.

File: src/x_misc.c (hunk 2 above)
Comment (porres) [reply src/x_misc.c:395]: what should I do here?

Comment (Spacechild1) [reply src/x_misc.c:395]: The lines starting at 395 seem to have 2 extra spaces of indentation.

**Comment by Spacechild1** [comment]
> I thought '-f' would be confusing as it is used for [oscformat] but it has a different meaning there ('format').

I see. What about `-n` for "numeric address pattern"?

> I also think there could be a message method to change this behaviour as well.

I don't think that's necessary. After all, this feature is really a bug fix in disguise to maintain backwards compatibility. I don't think that users would need to switch dynamically between the default "bad" mode and the new "good" mode. Otherwise, they can always use two objects.

**Comment by porres** [comment]
> What about -n for "numeric address pattern"?

sounds good, and what's your reasoning that this is better than an argument, it's more explicit?

**Comment by Spacechild1** [comment]
I think flags are more user-friendly than numeric codes. Also, we might want to add other options in the future and positional arguments don't scale well.

**Comment by umlaeute** [comment]
i was going to comment on the `-float` as well.
just for the record: my original suggestion would have been `-fp` (for: "**f**loat (in the)  **p**ath"), but `-n` looks ok as well.

**Comment by umlaeute** [comment]
> After all, this feature is really a bug fix in disguise to maintain backwards compatibility.

i do not agree with that statement. my reasoning is explained in the help patch for `[file split]` (which faces a similar dilemma).
the short version is: what do you do with `/foo/02/bar` *and* `/foo/2/bar`?

I'm kind of happy with a flag that enables the output as a number (as suggested here).
(it *might* be cool, if the actual flag used could also be used for `[file split]`.)

while we are there: how about adding a 2nd outlet to `[oscparse]` that gives the index in the output list that splits the address from the data?

**Comment by Spacechild1** [comment]
> while we are there: how about adding a 2nd outlet to [oscparse] that gives the index in the output list that splits the address from the data?

are you reading my mind? I was about to propose exactly the same thing, but then thought I should open a seperate feature request.

**Comment by umlaeute** [comment]
> are you reading my mind?

actually it's writing, but that ambiguity is all part of the bigger picture.

**Comment by reduzent** [comment]
> while we are there: how about adding a 2nd outlet to `[oscparse]` that gives the index in the output list that splits the address from the data?

Very nice. That missing info prevented me from switching to [oscformat]/[oscparse] in some projects. It didn't occur to me that it could be added so easily.

**Comment by Spacechild1** [comment]
@porres Can you update the PR title and description? Maybe just something like "[oscparse] improvements" and then a description of both features?

**Comment by porres** [comment]
done

**Comment by porres** [comment]
yay, thanks for merging this!

**Pull request merged**